documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+`IMotor` is defined as follows:
+```cs
+public interface IMotor
+{
+    float Speed { get; set; }
+}
+```
+
+This is exactly as in 2.X.
+
+**INFORMATIVE TEXT:** It is possible that, should integration with DualSense controllers be implemented, that more properties are added to `IMotor` than just `Speed`.
+
+`GamepadState` is defined as follows:
+```cs
+public struct GamepadState
+{
+    public IGamepad Device { get; init; }
+    public JoystickButtonState Buttons { get; init; }
+    public DualReadOnlyList<Vector2> Thumbsticks { get; init; }
+    public DualReadOnlyList<float> Triggers { get; init; }
+    public IReadOnlyList<IMotor> VibrationMotors { get; } // forwards to forwards Device.VibrationMotors for ease of use
+}
+```
+
+`GamepadState` reuses a lot of the joystick API types, which are defined later in this proposal.
+
+`Thumbsticks` contain the two thumbsticks on this gamepad. The X and Y values within this list range from -1 to 1: -1 being leftmost, and 1 being rightmost.
+
+`Triggers` contains the two triggers on thsi gamepad. The values within this list range from 0 to 1: 0 being unpressed, and 1 being fully pressed.
+
+Note the use of the `DualReadOnlyList` type. This is basically just:
+```cs
+public readonly struct DualReadOnlyList<T> : IReadOnlyList<T>
+{
+    public readonly T E0;
+    public readonly T E1;
+}
+```
+
+This is used where the list will only ever have exactly two elements, mainly because the "gamepad" form factor is standard and it doesn't make sense to have multiple thumbsticks or triggers given a human only has two thumbs or index fingers. More exotic devices should be exposed using the joystick API.
+
+# Joystick Input
+
+This is the polyglot interface for any other human input device that roughly meets the description of being "joystick".
+
+```cs
+public interface IJoystick : IInputDevice
+{
+    JoystickState State { get; }
+}
+```
+```cs
+public struct JoystickState
+{
+    public IJoystick Device { get; init; }
+    public InputReadOnlyList<float> Axes { get; init; }
+    public JoystickButtonState Buttons { get; init; }
+    public InputReadOnlyList<Vector2> Hats { get; init; }
+}
+```
+
+This is pretty closely modeled as in 2.X: `Axes` containing the individual axes that can be represented by this joystick, `Buttons` containing the buttons which can be pressed, and `Hats` containing the hats.
+
+**INFORMATIVE TEXT:** The only difference is `Hats` is now a `Vector2` instead of a `Position2D`. It is still intended that the X and Y values are only ever `0` or `1`, but this is not a requirement for more exotic backends.
+
+`JoystickButtonState` is defined as follows:
+```cs
+public struct JoystickButtonState
+{
+    public bool this[JoystickButton btn] { get; }
+    public InputReadOnlyList<JoystickButton> Down { get; init; }
+    public InputReadOnlyList<JoystickButton> Up { get; }
```

Comment:
Same init questions

Reply:
Addressed on line 670, doesn't make sense to be done any other way etc etc.